File: java/wpalermo/src/main/java/com/lejour/wpalermo/DesafioSilhuetas.java
package com.lejour.wpalermo;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * 
 * @author william.palermo
 *
 */


public class DesafioSilhuetas {

	final static String RESOURCE_PATH = "./src/main/resources/casos/";
	final static String FILE_NAME = "caso1.txt";
	final static String RESULT_FILE = FILE_NAME.split("\\.")[0] + "_RESULTADO.txt";

	private static Integer qtdCasos = 0;
	private static Integer linhasCount = 0;
	private static Stream<String> lines;

	public static void main(String[] args) throws IOException {

		List<List<Integer>> values = new ArrayList<List<Integer>>();

		Path readPath = Paths.get(RESOURCE_PATH + FILE_NAME);
		
		lines = Files.lines(readPath);

		
		List<String> casos = lines.map(String::valueOf).collect(Collectors.toList());
		
		qtdCasos = Integer.valueOf(casos.remove(0));

		casos.forEach(line -> {

			linhasCount++;

			if (linhasCount % 2 != 1) {
				List<Integer> list = new ArrayList<Integer>();
				for (String s : line.split(" "))
					list.add(Integer.valueOf(s));
				values.add(list);
			}

		});

		if (linhasCount / 2 != qtdCasos)
			System.out.println("WARN - Quantidade de linhas diferente da quantidade passada no header");

		List<String> returnable = new ArrayList<String>();

		for (List<Integer> lista : values) {

			Integer valorParcial = 0;
			Integer indexMaior = 0;
			Integer indexSegundoMaior = 0;
			Integer valor = 0;

			for (int i = 0; i < lista.size(); i++) {

				valor = lista.get(i);

				indexMaior = buscarIndexMaior(lista, i);

				if (indexMaior == null) {
					if (i + 1 == lista.size())
						break;

					indexSegundoMaior = buscarIndexSegundoMaior(lista, i);
					valor = lista.get(indexSegundoMaior);

					if (indexSegundoMaior == (i + 1))
						continue;
					else {

						for (Integer j : lista.subList(i + 1, indexSegundoMaior))
							valorParcial += valor - j;

						//System.out.println(valorParcial);
						i = indexSegundoMaior - 1;

					}

				} else if (indexMaior == (i + 1))
					continue;
				else {
					for (Integer j : lista.subList(i + 1, indexMaior))
						valorParcial += valor - j;

					//System.out.println(valorParcial);
					i = indexMaior - 1;
				}

			}

			returnable.add(valorParcial.toString());

		}

		//System.out.println("-------RESPOSTA------");
		//returnable.forEach(System.out::println);
		
		Files.write(Paths.get(RESOURCE_PATH + RESULT_FILE), returnable);
	}

	
	/**
	 * Metodo procura na lista se existe alguma outra coluna maior do que a atual.
	 * E verifica se essa coluna vem depois da atual na seguencia. 
	 * @param lista
	 * @param index
	 * @return
	 */
	private static Integer buscarIndexMaior(List<Integer> lista, Integer index) {

		Integer count = 0;
		Integer numero = lista.get(index);

		for (Integer i : lista) {
			if (i > numero && index < count)
				return count;
			count++;
		}

		return null;

	}

	/**
	 * No caso de ja ter encontrado a maior coluna, esse metodo eh responsavel por encontrar a segunda maior. 
	 * A partir desse momento, a segunda maior passa a ser considerada como a maior e a logica segue a mesma.
	 * @param lista
	 * @param index
	 * @return
	 */
	private static Integer buscarIndexSegundoMaior(List<Integer> lista, Integer index) {

		Integer count = index;
		Integer numero = lista.get(index);

		Integer segundoMaior = lista.subList(index + 1, lista.size()).stream().max(Integer::compare).get();

		//If para o caso de ter 2 numeros iguais seguidos e esse ser igual ao maior
		if (segundoMaior == numero) {
			count = index+1;
			for (Integer i : lista.subList(index + 1, lista.size())) {
				if (i == segundoMaior)
					return count;
				count++;
			}
		} else {

			for (Integer i : lista.subList(index, lista.size())) {
				if (i == segundoMaior)
					return count;
				count++;
			}
		}

		return null;

	}
}

File: node/desafio-silhuetas/app/silhuetas/desafioSilhuetas.js
'use strict'

const fileRead = require('../fileServices/fileRead.js');

exports.run = function(){
    fileRead.readFile(desafio);
   //desafio(null, 1);
};


var desafio = function(err, data){
    if(err){
        console.error(err.message);
        throw err;
    }

    console.log('Iniciando processamento das silhuetas' + data.length );

    var returnable = [];
    const casos = [];
    
    const qtdLinhas = data.shift();
    console.log('HEADER do arquivo: ' + qtdLinhas );
    console.log('Quantidade de casos: ' + data.length/2);

    if(data.length/2 != qtdLinhas)
        console.log('WARN - quantidade de linhas do arquivo diferente da quantidade passada no HEADER');
    else
        console.log('INFO - Quantidade de linhas OK');


    data.forEach((element, index) => {

        if(index %2 == 1){
            let linha = element.split(' ');

             
            
            
            casos.push(linha);

        }

    });

    var indexLinha = 0;

    for(indexLinha; indexLinha < casos.length; indexLinha++) {

        var linha = casos[indexLinha];
        let valor = 0;
        let valorParcial = 0;
        var index = 0;

        

        if(linha[linha.length-1] == '')
            linha = linha.slice(0, linha.length-1);

        for(index; index < linha.length; index++) {

            valor = linha[index];
            var indexMaior = buscarIndexMaior(linha, index);



            if(indexMaior == null){
                if(index + 1 == linha.length)
                    break;

                var indexSegundoMaior = buscarSegundoMaior(linha, index);
                valor = linha[indexSegundoMaior];

                if(indexSegundoMaior == (index + 1))
                    continue;
                else{

                    let linha1 = linha.slice(index + 1, indexSegundoMaior);

                    for(var element in linha1)
                        valorParcial = valorParcial + (valor - linha1[element]);

                    index = indexSegundoMaior - 1;
               
                }

            }else if(indexMaior == (index+1)){
                continue;
            }else{
                let linha1 = linha.slice(index +1, indexMaior);

                for(var element in linha1)
                    valorParcial = valorParcial + (valor - linha1[element]);
                

                index = indexMaior - 1;
            }


        };
            console.log(valorParcial);
            returnable.push(valorParcial);
            

     

    };
    

}


function buscarIndexMaior(lista, index){

    let numero = parseInt(lista[index]);

    var i=0;
    for(i; i<lista.length; i++){
        var element = parseInt(lista[i]);
        if(element > numero && index < i)
            return i;
    };

    return null;

};


function buscarSegundoMaior(lista, index){

    let count = index;
    let segundoMaior = 0
    let listAux = lista.slice(index+1, lista.length);
    if(index + 1  == lista.length-1)
        segundoMaior = lista[index+1]
    else
        segundoMaior = Math.max.apply(Math, listAux);




    if(segundoMaior == parseInt(lista[index])){

        var subList = lista.slice(index+1, lista.length);
        count = index+1;

        var i=0;
        for (var i in subList){
            let element = subList[i];
            if(element == segundoMaior)
                return count;
            
            count++;
        }

    }else{
        var subList = lista.slice(index, lista.length);

        for(var i in subList){
            let element = subList[i];
            if(element == segundoMaior)
                return count;

            count = count + 1;
        }
    };

    return null;
};